File: plugins/pitch/harmonizer/harmonizer.js
// Harmonizer — two parallel granular pitch shifters tuned to musical
// intervals (default: minor third + perfect fifth, a classic Stevie
// Wonder / Eventide harmonizer setting) mixed under the dry voice. Drop
// it on a lead vocal for instant three-part harmony.

function HarmonizerProcessor() {
  // At ratio=2 (max +12 semis), each grain reads 2*grainSize past samples,
  // so buffer must cover at least that. 200 ms is enough for ±octave.
  var bufLen = Math.ceil(sampleRate * 0.2);
  this.buf = new Array(bufLen);
  for (var k = 0; k < bufLen; k++) this.buf[k] = 0;
  this.bufLen = bufLen;
  this.write = 0;
  this.grainSize = Math.floor(sampleRate * 0.08);
  this.phase1 = 0;
  this.phase2 = Math.floor(this.grainSize * 0.33); // start offset so the
                                                    // two voices don't both
                                                    // hit their seams together
}

HarmonizerProcessor.parameterDescriptors = [
  { name: 'voice1', label: 'V1 semi',  defaultValue: 3,   minValue: -12, maxValue: 12 },
  { name: 'voice2', label: 'V2 semi',  defaultValue: 7,   minValue: -12, maxValue: 12 },
  { name: 'spread', label: 'Spread',   defaultValue: 0.7, minValue: 0,   maxValue: 1 },
  { name: 'mix',    label: 'Mix',      defaultValue: 0.5, minValue: 0,   maxValue: 1 }
];

function readGrain(buf, write, bufLen, phase, ratio, grainSize) {
  var r = write - grainSize * ratio + phase * ratio;
  while (r < 0) r += bufLen;
  while (r >= bufLen) r -= bufLen;
  var i = Math.floor(r), f = r - i;
  var j = i + 1; if (j >= bufLen) j = 0;
  var env = 0.5 - 0.5 * Math.cos(2 * Math.PI * phase / grainSize);
  return (buf[i] * (1 - f) + buf[j] * f) * env;
}

HarmonizerProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var v1 = parameters.voice1[0];
  var v2 = parameters.voice2[0];
  var spread = parameters.spread[0];
  var mix = parameters.mix[0];
  var r1 = Math.pow(2, v1 / 12);
  var r2 = Math.pow(2, v2 / 12);
  var grainSize = this.grainSize;
  var halfGrain = grainSize * 0.5;

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    this.buf[this.write] = x;

    // Each voice = two grain reads half-window apart, summed.
    var pA1 = this.phase1;
    var pB1 = (this.phase1 + halfGrain) % grainSize;
    var pA2 = this.phase2;
    var pB2 = (this.phase2 + halfGrain) % grainSize;
    var voice1 = readGrain(this.buf, this.write, this.bufLen, pA1, r1, grainSize)
               + readGrain(this.buf, this.write, this.bufLen, pB1, r1, grainSize);
    var voice2 = readGrain(this.buf, this.write, this.bufLen, pA2, r2, grainSize)
               + readGrain(this.buf, this.write, this.bufLen, pB2, r2, grainSize);

    var wet = (voice1 + voice2) * 0.5 * spread;

    this.phase1++; if (this.phase1 >= grainSize) this.phase1 = 0;
    this.phase2++; if (this.phase2 >= grainSize) this.phase2 = 0;
    this.write++;  if (this.write  >= this.bufLen)  this.write  = 0;

    output[i] = x * (1 - mix * 0.5) + wet * mix;
  }
  return true;
};

registerProcessor('harmonizer', HarmonizerProcessor);

File: plugins/pitch/harmonizer/Harmonizer.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Harmonizer (native port) — direct translation of harmonizer.js. Two
// parallel granular pitch shifters at musically-spaced intervals, summed
// under the dry voice. The JS version did 4 grain reads × per-sample cos
// = ~16 trig ops per sample, which was the most expensive plugin in this
// registry after shimmer-reverb. Hann LUT + JVM speed makes it free.
public final class Harmonizer implements VocalMonitorNativePlugin {

    private float[] buf;
    private int bufLen;
    private int write;
    private int grainSize;
    private int phase1;
    private int phase2;
    private float[] hann;

    private float voice1 = 3f, voice2 = 7f, spread = 0.7f, mix = 0.5f;

    @Override
    public void init(int sampleRate) {
        bufLen = (int) Math.ceil(sampleRate * 0.2);
        buf = new float[bufLen];
        write = 0;
        grainSize = (int) Math.floor(sampleRate * 0.08);
        phase1 = 0;
        phase2 = grainSize / 3;
        hann = new float[grainSize];
        for (int i = 0; i < grainSize; i++) {
            hann[i] = (float) (0.5 - 0.5 * Math.cos(2.0 * Math.PI * i / grainSize));
        }
    }

    @Override
    public String[] parameterNames() {
        return new String[] { "voice1", "voice2", "spread", "mix" };
    }

    @Override
    public float parameterMin(String name) {
        if ("voice1".equals(name) || "voice2".equals(name)) return -12f;
        return 0f;
    }

    @Override
    public float parameterMax(String name) {
        if ("voice1".equals(name) || "voice2".equals(name)) return 12f;
        return 1f;
    }

    @Override
    public float parameterDefault(String name) {
        switch (name) {
            case "voice1": return 3f;
            case "voice2": return 7f;
            case "spread": return 0.7f;
            case "mix":    return 0.5f;
            default:       return 0f;
        }
    }

    @Override
    public String parameterLabel(String name) {
        switch (name) {
            case "voice1": return "V1 semi";
            case "voice2": return "V2 semi";
            case "spread": return "Spread";
            case "mix":    return "Mix";
            default:       return name;
        }
    }

    @Override
    public void setParameter(String name, float value) {
        switch (name) {
            case "voice1": voice1 = value; break;
            case "voice2": voice2 = value; break;
            case "spread": spread = value; break;
            case "mix":    mix = value; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float r1 = (float) Math.pow(2.0, voice1 / 12.0);
        final float r2 = (float) Math.pow(2.0, voice2 / 12.0);
        final int gs = grainSize;
        final int halfGs = gs / 2;
        final int bL = bufLen;
        final float[] hannLocal = hann;
        final float[] bufLocal = buf;
        final float spreadScaled = spread * 0.5f;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal * 0.5f;
        int w = write;
        int p1 = phase1;
        int p2 = phase2;

        final int n = input.length;
        for (int i = 0; i < n; i++) {
            final float x = input[i];
            bufLocal[w] = x;

            // Voice 1, two grains.
            float v1 = readGrainSum(bufLocal, w, bL, p1, halfGs, gs, r1, hannLocal);
            // Voice 2.
            float v2 = readGrainSum(bufLocal, w, bL, p2, halfGs, gs, r2, hannLocal);

            float wet = (v1 + v2) * spreadScaled;

            p1++; if (p1 >= gs) p1 = 0;
            p2++; if (p2 >= gs) p2 = 0;
            w++;  if (w  >= bL) w  = 0;

            output[i] = x * dry + wet * mixLocal;
        }

        write = w;
        phase1 = p1;
        phase2 = p2;
    }

    // Two crossfaded grain reads at `ratio` speed, summed. Inlined into
    // the loop above would be marginally faster; pulling it out keeps the
    // bytecode smaller and lets the JIT spot the pattern.
    private static float readGrainSum(float[] buf, int write, int bufLen,
                                      int phase, int halfGs, int gs,
                                      float ratio, float[] hann) {
        int pA = phase;
        int pB = phase + halfGs;
        if (pB >= gs) pB -= gs;
        float rA = write - gs * ratio + pA * ratio;
        float rB = write - gs * ratio + pB * ratio;
        while (rA < 0) rA += bufLen;
        while (rA >= bufLen) rA -= bufLen;
        while (rB < 0) rB += bufLen;
        while (rB >= bufLen) rB -= bufLen;
        int iA = (int) rA; float fA = rA - iA;
        int jA = iA + 1; if (jA >= bufLen) jA = 0;
        float sA = buf[iA] * (1f - fA) + buf[jA] * fA;
        int iB = (int) rB; float fB = rB - iB;
        int jB = iB + 1; if (jB >= bufLen) jB = 0;
        float sB = buf[iB] * (1f - fB) + buf[jB] * fB;
        return sA * hann[pA] + sB * hann[pB];
    }
}
